lms/djangoapps/courseware/views/views.py

```diff
@@ -1338,12 +1338,24 @@ def _course_home_redirect_enabled():
 
 @login_required
 @ensure_valid_course_key
-def submission_history(request, course_id, student_username, location):
+def submission_history(request, course_id, student_username_or_email, location):
     """Render an HTML fragment (meant for inclusion elsewhere) that renders a
     history of all state changes made by this user for this problem location.
     Right now this only works for problems because that's all
     StudentModuleHistory records.
     """
+    user = None
+    found_user_name = None
+    if '@' in student_username_or_email:
```

Comment:
This would require two queries if `student_username_or_email` contains an email address. Try the following.

1. Avoid extra database hit. 
2. Its more cleaner 

```
from django.core.exceptions import ValidationError
from django.core.validators import validate_email

def validate_learner_identifier(learner_identifier):
    kwargs = {}
    try:
        validate_email(learner_identifier)
    except ValidationError as e:
        kwargs = {'username': learner_identifier}
    else:
        kwargs = {'email': learner_identifier}
    return kwargs


def get_learner_username(learner_identifier):
    kwargs = validate_learner_identifier(learner_identifier)
    learner = User.objects.filter(**kwargs)
    if learner.exists():
        return learner.first().username



def submission_history(request, course_id, student_username_or_email, location):
    ...
    ...
    ...
    ...
    learner_username = get_learner_username(learner_identifier) # or student_username both are readable.
    if not learner_username:
        return HttpResponse(escape(_('User does not exist.')))
    ...
    ...
    ...
```